### ISSUE TITLE
[Optimizer Perf] Improve speed of _init_group to c++

### ISSUE THREAD

**Issue by OrenLeung** [pytorch/pytorch#108090]
### 🚀 The feature, motivation and pitch

It has been noticed on many models that[ _init_group](https://github.com/pytorch/pytorch/blob/main/torch/optim/adamw.py#L90-L146) takes 33% of the whole `optimizer.step()` on **every step**  for popular optimizers like adam, adamw, etc. since it has to loop through each single param for each step

## Proposal, do this in c++

just like how `_group_tensors_by_device_and_dtype` in https://github.com/pytorch/pytorch/pull/100007 was moved in c++, lets move _init_group there too

cc: @janeyx99 @crcrpar

cc: @clchan **@ezhang887******

**Comment by albanD** [comment]
Can you share sample model/optimizer that shows this.
Doing a couple if statements and appends for each param shouldn't take 33% of the optimizer.step(). This sounds very surprising.

**Comment by tringwald** [comment]
I just ran a [small toy network](https://gist.github.com/tringwald/8559a5a498089eeb08f66abf1e8ef2c3), which I modified to have an additional 1000 linear layers. The results indeed show that `_init_group` takes a lot of the overall processing time. The image below shows the top few function calls for a simple training run with `Adam`.

![grafik](https://github.com/pytorch/pytorch/assets/81091441/98da6dbb-dc9f-403f-a147-798b7b14058d)

**Comment by janeyx99** [comment]
Currently, our _init_group() handles:
- initializing the state at the first step call. This should only happen once!
- creating the lists of parameters and state with grads that we will pass into the step to avoid compute when for when grads are None

The state initialization taking a bit (for the first step) is expected, but I don't think the Python list creation should take so long/will be much faster in C++. It would be interesting to have numbers proving otherwise.

Another reason that may be relevant here is if the actual computation is small (like in your small toy example), versus having lots of big parameters to fuse together. I'd anticipate seeing the percentage change depending on model size.

**Comment by tringwald** [comment]
I just repeated the experiment on `2.1.0`:

- With an additional 10 linear layers of size 1024x1024, `_init_group` amounted to ~1.2% of total time
- With an additional 1000 linear layers of size 16x16, `_init_group` amounted to ~4.7% of total time (similar to my earlier post)

While runtime is definitely more reasonable with fewer layers, `_init_group` still took longer than the 350,000 calls to `normalize` in the batch preprocessing.
